ODIN_II/SRC/include/HardSoftLogicMixer.hpp

```diff
@@ -0,0 +1,107 @@
+    void map_deferred_blocks(netlist_t* netlist);
+
+    /*----------------------------------------------------------------------
+     * Function: note_candidate_node
+     * Calculates number of available hard blocks by issuing a call,
+     * traverses the netlist and statistics to figure out
+     * which operation should be implemented on the hard block
+     * Parameters: 
+     *      node_t * : pointer to candidate node
+     *      operation_list : type of the hard block to optimize
+     * returns:
+     *---------------------------------------------------------------------*/
+    void note_candidate_node(nnode_t* node, operation_list type);
+
+  private:
+    /* ----------------------------------------------------------------------
+     * Function parse_opt_parameters
+     * This function parses the configuration bitmask and sets
+     * the corresponding boolean variables in the _enabledOptimizationsArray.
+     *----------------------------------------------------------------------*/
+    void parse_opt_parameters(const config_t);
+
+    /*----------------------------------------------------------------------
+     * Function: grid_statistics
+     *   For all the layouts represented in the architecture file, populates
+     *   array of all optimizations with estimated numbers for hard blocks
+     *---------------------------------------------------------------------*/
+    void grid_statistics();
+
+    /* ----------------------------------------------------------------------
+     * Function: soft_map_remaining_nodes
+     * After finishing mapping logic nodes onto hard blocks, 
+     * this function should be called to assure remaining blocks
+     * are assigned to be implemented in soft logic
+     *---------------------------------------------------------------------*/
+    void soft_map_remaining_nodes(netlist_t* netlist);
+
+    /*----------------------------------------------------------------------
+     * Function: hard_blocks_needed
+     * Returns cached value calculated from netlist, for a specific optimiza
+     * tion kind
+     *---------------------------------------------------------------------*/
+    int hard_blocks_needed(operation_list);
+
+    /*----------------------------------------------------------------------
+     * Function: choose_hard_blocks
+     * Functiion selects and instantiates operations on hard blocks for a 
+     * given optimization kind
+     *---------------------------------------------------------------------*/
+    void choose_hard_blocks(netlist_t* netlist);
+
+    // This array is composed of vectors, that store nodes that
+    // are potential candidates for performing mixing optimization
+    std::vector<nnode_t*> _nodes_by_opt[operation_list::operation_list_END];
+
+    // These booleans store devices selected for optimization
+    bool _allOptsDisabled;
+    // This is a container containing all optimization passes
+    std::vector<MixingOpt*> _enabled_opts;
```

Comment:
change to MixingOpt* _enabled_opts[operation_list_END];

Reply:
Fixed, thank you

Reply:
Done, thanks for the comment!